File: server/app/routes/ws_console.py
from __future__ import annotations

from typing import Any

from server.app.protocol import envelope, receive_json, send_json
from server.app.runtime import sock
from server.app.services.events import console_subscriptions, send_chat_snapshots, send_node_update, send_payload


@sock.route("/ws/console")
def console_socket(ws: Any) -> None:
    subscription_id: int | None = None
    try:
        while True:
            message = receive_json(ws)
            if message.get("type") != "console.subscribe":
                payload = envelope(
                    "server.notice",
                    {"level": "warning", "message": "expected console.subscribe"},
                    source_kind="server",
                    source_id="main",
                    target_kind="console",
                    target_id="browser",
                    request_id=message.get("request_id"),
                )
                if subscription_id is None:
                    send_json(ws, payload)
                else:
                    send_payload(subscription_id, payload)
                continue

            chat_ids = message.get("data", {}).get("chat_ids", [])
            watch_nodes = bool(message.get("data", {}).get("watch_nodes", False))

            if subscription_id is None:
                subscription_id = console_subscriptions.add(
                    ws,
                    chat_ids,
                    watch_nodes=watch_nodes,
                )
            else:
                console_subscriptions.update(
                    subscription_id,
                    chat_ids,
                    watch_nodes=watch_nodes,
                )

            send_chat_snapshots(subscription_id, message.get("request_id"))
            if watch_nodes:
                send_node_update(subscription_id, message.get("request_id"))
    except ConnectionError:
        if subscription_id is not None:
            console_subscriptions.remove(subscription_id)

File: cmd/qunce/ws_console.go
package main

import (
	"log"
	"net/http"
	"strings"

	"github.com/gorilla/websocket"
)

var wsConsoleUpgrader = websocket.Upgrader{
	CheckOrigin: func(_ *http.Request) bool {
		return true
	},
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

func registerConsoleSocket(mux *http.ServeMux, consoles *consoleRegistry) {
	mux.HandleFunc("/ws/console", handleConsoleSocket(consoles))
}

func handleConsoleSocket(consoles *consoleRegistry) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		conn, err := wsConsoleUpgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Printf("upgrade console socket: %v", err)
			return
		}

		var subscriptionID int64
		defer func() {
			if subscriptionID > 0 {
				consoles.remove(subscriptionID)
			}
			_ = conn.Close()
		}()

		for {
			raw, err := readEnvelope(conn)
			if err != nil {
				return
			}
			payload, err := dataMap(raw.Data)
			if err != nil {
				_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
					"server.notice",
					"server",
					"main",
					"console",
					"browser",
					map[string]interface{}{"level": "warning", "message": "invalid subscribe payload"},
					raw.RequestID,
				)))
				continue
			}

			if strings.TrimSpace(raw.Type) != "console.subscribe" {
				_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
					"server.notice",
					"server",
					"main",
					"console",
					"browser",
					map[string]interface{}{"level": "warning", "message": "expected console.subscribe"},
					raw.RequestID,
				)))
				continue
			}

			var chatIDs []string
			if rawChatIDs, ok := payload["chat_ids"].([]interface{}); ok {
				for _, rawChatID := range rawChatIDs {
					chatID, ok := rawChatID.(string)
					if !ok {
						continue
					}
					chatID = strings.TrimSpace(chatID)
					if chatID == "" {
						continue
					}
					chatIDs = append(chatIDs, chatID)
				}
			}
			watchNode := toBool(payload["watch_nodes"], false)

			if subscriptionID == 0 {
				subscriptionID = consoles.add(conn, chatIDs, watchNode)
			} else {
				consoles.update(subscriptionID, chatIDs, watchNode)
			}

			consoles.sendChatSnapshot(subscriptionID, raw.RequestID, nil)
			if watchNode {
				consoles.sendNodeUpdate(subscriptionID, raw.RequestID, nil)
			}
		}
	}
}
